Add Azure Blob Storage as backend

diptadas opened the pull request:
### Environment variables:

```

AZURE_ACCOUNT_NAME=storage-account-name
AZURE_ACCOUNT_KEY=storage-account-key

```

### Environment variables for test:

```

RESTIC_TEST_AZURE_ACCOUNT_NAME=storage-account-name
RESTIC_TEST_AZURE_ACCOUNT_KEY=storage-account-key
RESTIC_TEST_AZURE_REPOSITORY=azure:restic-test-container

```

### Init repository:

```

$ restic -r azure:container-name/prefix/dir init
or
$ restic -r azure://container-name/prefix/dir init

```
Container location is fixed for a azure storage account. So we cant specify it in repository.

tamalsaha commented:
@fd0 , you will need a Azure storage account to run these tests. If you would like us to provide the account, please share an email address. I can give you access to our test Azure test storage account that you will be able to use to run tests. You can find my email on my Github profile.

fd0 commented:
FYI, I don't have much time at the moment, will check this later. Thanks for the contribution!

diptadas commented:
**Format:**
`$ restic -r azure:container-name/prefix/directory init`

tamalsaha commented:
@diptadas , we should change the URL to `azure:container:/[prefix]` .  There should be a `:` after container name.

diptadas commented:
**Format:** `azure:container:/[prefix] `

tamalsaha commented:
Our pr for reading partial blobs from Azure storage is now merged into the official sdk repo. https://github.com/Azure/azure-sdk-for-go/pull/664

flo82 commented:
I'm backing up to minio running in gateway mode. gateway mode supports azure storage. see documentation [here](https://github.com/minio/minio/tree/master/docs/gateway).
it would be interesting which method is faster and more stable:
a) restic -> minio -> azure
b) restic -> azure

tamalsaha commented:
@flo82 , if you get a chance to run benchmarks, please share with us. It is kind of hard to see how an extra hop will perform any better.

flo82 commented:
i will try, if i find time for that :-)
will it be possible to switch between a) and b) on the fly (because minio impl and azure impl use the same object/file structure) or do i need to init a new repository?

tamalsaha commented:
> will it be possible to switch between a) and b) on the fly (because minio impl and azure impl use the same object/file structure) or do i need to init a new repository?

It really depends on what Minio gateway does. It minio gateway maintains 1:1 mapping, it should be possible.

fd0 commented:
@diptadas @tamalsaha Hey, I'll test this in the next couple of days, is it okay for you if I squash the commits and add the necessary code to merge this? Or do you want to do this yourself, and I'll just add comments? I think the former will be much faster, but I'd like your approval before I do this. In order to build/verify the integration tests this needs to be reopened from a branch of the restic repo anyway.

diptadas committed:
Add Azure Blob Storage as backend

Environment variables:
AZURE_ACCOUNT_NAME=storage-account-name
AZURE_ACCOUNT_KEY=storage-account-key

Environment variables for test:
RESTIC_TEST_AZURE_ACCOUNT_NAME=storage-account-name
RESTIC_TEST_AZURE_ACCOUNT_KEY=storage-account-key
RESTIC_TEST_AZURE_REPOSITORY=azure:restic-test-container

Init repository:
$ restic -r azure:container-name:/prefix/dir init

tamalsaha commented:
@fd0, both the prs are squashed. We are happy with you adding necessary code to merge this into master. But we would like you to cherry-pick the squashed commit from our repo so that our contribution is visible in the restic master eventually. As @diptadas employer, I also speak for himself on this matter.

fd0 commented:
Awesome, then I'll do that. Cherry-picking your commits was the plan, I should've made that clearer. I've done the same for the Swift backend (see https://github.com/restic/restic/pull/975/commits).

fd0 commented:
Superseded by #1149
